### ISSUE TITLE
0.5.0-rc2 Up to 2x performance regression loading precompiled packages

### ISSUE THREAD

**Issue by ufechner7** [JuliaLang/julia#18030]
Julia 0.4.6
```julia
tic(); using PyPlot; toc()
elapsed time: 2.569278583 seconds
```
Julia 0.5.0-rc2
```julia
tic(); using PyPlot; toc();
elapsed time: 5.360103935 seconds
```
This happens with precompiled packages, therefore - to my understanding - the root cause cannot be, that the new llvm version is slower compiling.
It would be very nice, if this regression could be fixed.
Computer: Linux 64 bit, Ubuntu 14.04,   i7-3770 CPU @ 3.40GHz × 4

**Comment by yuyichao** [comment]
Precompile doesn't remove codegen overhead. Although it can't explain all the difference, it seems that at least 50% of the time is indeed spent in llvm.

**Comment by ufechner7** [comment]
Some more results. They are not consistent: Loading Gtk (master branch) is even faster on 0.5, compared to 0.4.6. The slow down factor of loading JuMP is 1.5 .

```julia
Julia 0.4.6
tic(); using PyPlot; toc()
elapsed time: 2.569278583 seconds

julia> tic(); using Gtk; toc();
elapsed time: 1.424508958 seconds

julia> tic(); using JuMP; toc();
elapsed time: 0.617230441 seconds


Julia 0.5.0-rc2

tic(); using PyPlot; toc();
elapsed time: 5.360103935 seconds

julia> tic(); using Gtk; toc();
elapsed time: 0.943486983 seconds

julia> tic(); using JuMP; toc();
elapsed time: 0.910472513 seconds
```

**Comment by ufechner7** [comment]
Why is precompilation not removing the codegen overhead? I thought, that this is exactly the purpose of precompilation?

**Comment by tkelman** [comment]
we don't currently save native code in the .ji files, so llvm still has work to do

**Comment by ufechner7** [comment]
Witch data format is stored in the .jl files?

**Comment by yuyichao** [comment]
Serialized (inferred) AST.

**Comment by ufechner7** [comment]
Why does it have so many items now?

**Comment by ViralBShah** [comment]
From a user perspective, it would be nice if PyPlot loading time is as before or lesser. I hope there is something we can do here.

**Comment by stevengj** [comment]
@ufechner7, in 0.5, every function corresponds to a unique type.

**Comment by ufechner7** [comment]
Is there a way to determine the number of unique types, that are defined?

**Comment by JeffBezanson** [comment]
Most of the types that exist in the system are not explicitly defined, but derived --- mostly tuples of various combinations of types actually. It's possible to get counts by poking into the system, but I'm not really sure how it would help to know.

**Comment by ufechner7** [comment]
Well, if the performance of jl_recache_types for large number of types is the bottleneck, it would be good to know the number in different scenarios. For example, is this number after loading PyPlot really so much higher then after loading Gtk.

**Comment by stevengj** [comment]
Maybe just add `printf("flagref_list.len = %zd", flagref_list.len);` in `jl_recache_types`?

**Comment by ufechner7** [comment]
Same results with 0.5-rc3 .

**Comment by vtjnash** [comment]
fixed by #18191. now there should only be dozens of types to recache.

**Comment by vtjnash** [comment]
fwiw, I blame @carnaval for this regression. :P
